### PR TITLE
CDAP-4199 Add Kitematic information to Docker documentation

### PR DESCRIPTION
**Pull request by awholegunch**
Fix for https://issues.cask.co/browse/CDAP-4199. Adds details on using Docker Kitematic to start a CDAP instance. 

Kitematic is Docker's GUI for Mac OS X and Windows for downloading containers, starting and managing them. Kitematic is in beta on the Mac (works, and the CDAP container works), and alpha on Windows (doesn't work; launching Kitematic fails under Windows 8 under Parallels; don't get to testing of the CDAP container).

Passes [Quick Build](http://builds.cask.co/browse/CDAP-DOB51-4).

[Docker page](http://builds.cask.co/artifact/CDAP-DOB51/shared/build-4/Docs-HTML/3.3.0-SNAPSHOT/en/developers-manual/getting-started/standalone/docker.html)

[Updated to Build 4.]

### CODE REVIEW

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 1]
```diff
@@ -3,15 +3,85 @@
...
+It is a graphical user interface for running Docker containers. Follow these steps to install 
+Kitematic and then download, start, and connect to an instance of CDAP.
+
+#. Download and install the `Docker Toolbox <https://www.docker.com/docker-toolbox>`__ for 
+   either Mac OS X or Microsoft Windows.
+
+#. Start Kitematic. On Mac OS X, it will be installed in ``/Applications/Docker/Kitematic``; on 
+   Windows, in ``Start Menu > Docker > Kitematic``.
+   
+#. Once Kitematic has started |---| you can either login to the Docker Hub or skip that step |---|
```
Comment (jwang47) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: I think `you can either login to the Docker Hub or skip that step` can be omitted, since it seems to be saying "you can either login or not login".

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 2]
```diff
@@ -3,15 +3,85 @@
...
+`Docker Kitematic <https://www.docker.com/docker-kitematic>`__ is available as part of the
+`Docker Toolbox <https://www.docker.com/docker-toolbox>`__ for either Mac OS X or Microsoft Windows.
+It is a graphical user interface for running Docker containers. Follow these steps to install 
+Kitematic and then download, start, and connect to an instance of CDAP.
+
+#. Download and install the `Docker Toolbox <https://www.docker.com/docker-toolbox>`__ for 
+   either Mac OS X or Microsoft Windows.
+
+#. Start Kitematic. On Mac OS X, it will be installed in ``/Applications/Docker/Kitematic``; on 
+   Windows, in ``Start Menu > Docker > Kitematic``.
```
Comment (jwang47) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Maybe we want to skip the previous 2 steps, as the URLs and paths may change without notice. Or, at least take out the URLs and paths. Also, I'd guess that most people who want to use CDAP on Docker already have experience using Docker.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 3]
```diff
@@ -3,15 +3,85 @@
...
+
+#. Close the menu by pressing the ``X`` in the circle. Press "Create" to download and start the CDAP image. 
+   When it has started up, you will see in the logs a message that the CDAP UI is listening on port 9999:
+ 
+     .. image:: ../../_images/kitematic/kitematic-4-cdap-started.png
+        :width: 8in
+        :align: center
+
+#. To connect a web browser for the CDAP UI, you'll need to find the IP address that the 
+   Kitematic instance is proxying the CDAP instance. The easiest way to do that is click
```
Comment (jwang47) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Isn't it the Docker instance rather than the Kitematic instance? From these docs it looks like Kitematic is just an UI for Docker.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 4]
```diff
@@ -3,15 +3,85 @@
...
+        :width: 8in
+        :align: center
+
+#. Click on the tags button:
+ 
+     .. image:: ../../_images/kitematic/kitematic-2-tags.png
+        :width: 4in
+        :align: center
+
+#. Scroll the window to the last version shown, and select it (unless you'd like a different version).
```
Comment (jwang47) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: How about "Select the desired version." instead?

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 2 above)
Comment (jwang47) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Also, if Docker or Kitematic change their URLs and paths, docs for older versions will be broken. So I think we generally shouldn't include variables from outside systems.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 1 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: The reason I added it was because Docker Kitematic does not make it clear that you can skip that step. They make it look like you need to, but turns out everything works if you skip it. I can take it out, or rewrite it so that it's clearer..

Comment (jwang47) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Personally, after looking at the Kitematic images here, I assumed that logging in was only required to access private repositories (?).

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 2 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: I 'd suggest that—at least on the Mac—this is a great way to get people into CDAP, especially Hydrator, for those **with no experience** in any of these technologies. As far as breaking paths, that can happen at any time; we can run (and do) run link checkers, but I think that the advantages of having links to other places out weighs broken links. Are you suggesting we remove **all** external links?

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 4 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: The reason is that people might select the tag "latest", thinking that they were going to get the "latest" as in the most-current, most-up-to-date. Turns out that for Docker, "latest" seems to mean "last-one-that-Chris-installed-at-Docker-hub", which could be any version. 

Someone (like someone doing testing or experimenting) might find this useful to pull down and run a particular version of CDAP. I can imagine I could be using this to quickly test things between different versions.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 3 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Well, I said it that way, because in the image of the UI, you see something labelled "Docker Port", which is internal to the VM, and something labelled "Mac.IP-Port", which is what is externally facing and what the browser needs to connect to. (Since I could not get the Windows Alpha to run, I don't know what the Windows version labels that as.)

Because of that, I didn't want to confuse the issue by saying "Docker" and have people think that they needed to use the ``9999`` port. I chose to use "Kitematic instance" to cover both Mac and Windows, and so they would associate that with the external facing side rather than the internal.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 2 above)
Comment (jwang47) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: I see, that makes sense. In that case, I think we should write a bit about Docker as an easier way to get into CDAP without having to manually setup anything.

For breaking paths, my main concern is having to maintain older versions of the documentation. For example, if Docker changes anything we documented here after CDAP 3.3, we'll have to apply patches to multiple versions of the docs. Furthermore, if we include the docs in the SDK/VM/Maven releases, then (1) the links will eventually break and/or (2) the docs will no longer apply without any way for us to handle it. In this case, we can't do anything about (2) unless we document a particular version of Kitematic and Docker, but we can do something about (1).

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 4 above)
Comment (jwang47) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: I wasn't commenting about the sentence talking about "latest" (comes after). Instead, I was making a comment to help minimize wordiness in the first sentence.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 3 above)
Comment (jwang47) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: But there is no "Kitematic instance", only a Docker instance, right? Maybe reword as "To use the CDAP UI, you'll need to find the external IP addresses and ports that the Docker instance is exposing."

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 5]
```diff
@@ -3,15 +3,85 @@
...
+   on the *Settings* tab, and then the *Ports* tab:
+ 
+     .. image:: ../../_images/kitematic/kitematic-5-links.png
+        :width: 8in
+        :align: center
+
+
+#. This shows that the CDAP instance is listening on Docker port ``9999``, and the Kitematic 
+   instance is listening on ``192.168.99.1100:32769``. The text in blue is a link;
+   clicking it will open it in your system web browser and connect to the CDAP UI:
```
Comment (jwang47) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Can say "This shows that the CDAP instance is listening on the internal port `9999` within the Docker instance, and exposes that port on the external IP address and port `192.168.99.1100:32769`.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 1 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Fixed by removing the text inside the em dashes.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 3 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Fixed.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 4 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Fixed.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 5 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Fixed.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 2 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Added text about Docker being an "easy to use CDAP" at the start of the page. 

We don't include the docs in the releases; they are only available either on the website or as a downloadable zip. But I think people understand—especially with a rapidly-released product like CDAP—that there will be issues with older documentation being stale.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst [hunk 6]
```diff
@@ -3,15 +3,88 @@
...
+        :width: 8in
+        :align: center
+
+#. Click on the tags button:
+ 
+     .. image:: ../../_images/kitematic/kitematic-2-tags.png
+        :width: 4in
+        :align: center
+
+#. Scroll the window to the last version shown, and select the desired version.
```
Comment (jwang47) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Just have `Select the desired version.`

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 2 above)
Comment (jwang47) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Sounds good.

File: cdap-docs/developers-manual/source/getting-started/standalone/docker.rst (hunk 6 above)
Comment (awholegunch) [inline cdap-docs/developers-manual/source/getting-started/standalone/docker.rst]: Fixed.

**Comment by jwang47** [comment]
LGTM
